pkg/frontend/adminactions/serialtool.go

```diff
@@ -0,0 +1,115 @@
+package adminactions
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	mgmtcompute "github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2019-03-01/compute"
+	mgmtstorage "github.com/Azure/azure-sdk-for-go/services/storage/mgmt/2019-04-01/storage"
+	azstorage "github.com/Azure/azure-sdk-for-go/storage"
+	"github.com/Azure/go-autorest/autorest/azure"
+	"github.com/Azure/go-autorest/autorest/date"
+	"github.com/sirupsen/logrus"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/mgmt/compute"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/mgmt/storage"
+	"github.com/Azure/ARO-RP/pkg/util/refreshable"
+	"github.com/Azure/ARO-RP/pkg/util/stringutils"
+)
+
+func (a *adminactions) SerialTool(ctx context.Context, w http.ResponseWriter, log *logrus.Entry,
+	fpAuth refreshable.Authorizer, r azure.Resource, oc *api.OpenShiftCluster, vmName string) error {
+	s, err := newSerialTool(log, compute.NewVirtualMachinesClient(r.SubscriptionID, fpAuth),
+		storage.NewAccountsClient(r.SubscriptionID, fpAuth), oc)
+	if err != nil {
+		return err
+	}
+
+	return s.dump(ctx, vmName, w)
+}
+
+type serialTool struct {
+	log *logrus.Entry
+	oc  *api.OpenShiftCluster
+
+	accounts        storage.AccountsClient
+	virtualMachines compute.VirtualMachinesClient
+
+	clusterResourceGroup string
+}
+
+func newSerialTool(log *logrus.Entry, virtualMachines compute.VirtualMachinesClient,
+	accounts storage.AccountsClient, oc *api.OpenShiftCluster) (*serialTool, error) {
+
+	return &serialTool{
+		log: log,
+		oc:  oc,
+
+		accounts:        accounts,
+		virtualMachines: virtualMachines,
+
+		clusterResourceGroup: stringutils.LastTokenByte(oc.Properties.ClusterProfile.ResourceGroupID, '/'),
+	}, nil
+}
```

Comment:
I think remove this struct and just have one function that creates the client and writes, it's not providing much purpose at this point

Reply:
Good point